### PR TITLE
Faster query string decoder

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

decodeParams is not accounting for common decoding scenarios
making it highly unpredictable.

Modifications:

Arrange control flow to isolated predictable from unpredictable
checks to improve CPU usage (stalled frontend cycles)

Result:

Faster query decoding

### CODE REVIEW

**Comment by franz1981** [comment]
@vietj too

**Comment by franz1981** [comment]
These are the numbers with these changes:
```
Benchmark                                                Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecoding               thrpt   40   3.196 ± 0.050  ops/us
QueryStringDecoderBenchmark.noDecoding                  thrpt   40  12.903 ± 0.096  ops/us
QueryStringDecoderBenchmark.nonStandardDecoding         thrpt   40   4.291 ± 0.103  ops/us
QueryStringDecoderBenchmark.onlyDecoding                thrpt   40   3.831 ± 0.011  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameter  thrpt   40  23.055 ± 0.251  ops/us
```
compared to before:
```
Benchmark                                                Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecoding               thrpt   40   2.641 ± 0.170  ops/us
QueryStringDecoderBenchmark.noDecoding                  thrpt   40  11.413 ± 0.112  ops/us
QueryStringDecoderBenchmark.nonStandardDecoding         thrpt   40   3.219 ± 0.058  ops/us
QueryStringDecoderBenchmark.onlyDecoding                thrpt   40   3.239 ± 0.114  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameter  thrpt   40  22.001 ± 0.404  ops/us
```

**Comment by franz1981** [comment]
@viet I've added a new inteface (bad name, to improve), which could save users which have to perform 
single shot decoding to be forced collecting parameters in the provided linked map (IIRC Vertx prefer a `MultiMap`).

wdyt?

@chrisvest I'm opened to ideas to provide a decent API here: impl wise I've preferred to do some "dirty trick" 
with parameters/return types to avoid copying the core algorithm twice.

Perf-wise these is what we get with the new API:
```
Benchmark                                                     Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecodingNoMap               thrpt   40   3.722 ± 0.070  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                  thrpt   40  20.763 ± 0.040  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap         thrpt   40   5.231 ± 0.017  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                thrpt   40   4.157 ± 0.019  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap  thrpt   40  38.208 ± 0.855  ops/us
```

Which remove the map allocation from the equation (that's unfair: users will likely have their own bespoke map there!)

**Comment by franz1981** [comment]
@chrisvest when the API and the PR is stable I'll write/modify the existing tests to use it.

**Comment by franz1981** [comment]
I Will perform some assembly analysis because some of the numbers are really weird and I suspect there is something going on inlining side.

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 1]
```diff
@@ -223,55 +218,176 @@ private int pathEndIdx() {
...
+
+    private static int pathEndIdx(URI uri) {
+        return uri.getRawPath().length();
+    }
+
+    public static int getDefaultMaxParams() {
+        return DEFAULT_MAX_PARAMS;
+    }
+
+    public static boolean decodeParams(URI uri, Charset charset, int maxParams, boolean semicolonIsNormalChar,
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:239]: These methods should have some javadocs to explain them.

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 2]
```diff
@@ -196,7 +191,7 @@ public String path() {
      */
     public Map<String, List<String>> parameters() {
         if (params == null) {
-            params = decodeParams(uri, pathEndIdx(), charset, maxParams, semicolonIsNormalChar);
+            params = decodeParams(uri, pathEndIdx(), charset, null, maxParams, semicolonIsNormalChar);
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:194]: We could also add a variant of this method that returned void and took the callback as a parameter.

Comment (franz1981) [reply codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:194]: That's indeed a very good idea!

**Comment by franz1981** [comment]
@chrisvest I didn't overloaded the static methods to perform parameter decoding yet: do you think it to be a nice addition?
I've supposed it wasn't a widely adopted feature, really, but useful, so didn't wanted to make it "heavy" by adding many different methods.

Test-wise, I've added an additional `assertParameters` sanity check for all existing tests which perform some form of parameter validation, but it is using the existing instance method only. 
I can add calling the other static ones, which perform some null checks,let me know your opinion here

**Comment by franz1981** [comment]
I have another improvement coming from the decodeComponent part, hold on!

**Comment by vietj** [comment]
It would be great if we could have a _static_ way of populating existing data structures, like 

```java
  public interface Populator<T> {

    void setQueryParam(String name, String value, T map);

  }

  public class MultimapPopulator implements Populator<Map<String, String>> {
    @Override
    public void setQueryParam(String name, String value, Map<String, String> map) {
      map.put(name, value);
    }
  }
```

**Comment by franz1981** [comment]
thanks @vietj for the feedback, it's a great idea indeed!

**Comment by franz1981** [comment]
After reading the different versions of the assembly produced by the benchmark I believe is it worthy to analyze what happen if the input sequence makes the JIT to decide for a different layout and order of the evaluated conditions (for the decoder state machine).

Hence, I've added a parameter to "confuse" JIT, and these are the numbers...
This PR:
```
Benchmark                                                       (confuseJIT)   Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecodingNoMap                         false  thrpt   40   3.531 ± 0.081  ops/us
QueryStringDecoderBenchmark.mixedDecodingNoMap                          true  thrpt   40   3.393 ± 0.043  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                            false  thrpt   40  21.158 ± 0.311  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                             true  thrpt   40  18.403 ± 0.460  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                   false  thrpt   40   5.163 ± 0.302  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                    true  thrpt   40   4.104 ± 0.101  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                          false  thrpt   40   4.192 ± 0.007  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                           true  thrpt   40   3.775 ± 0.068  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap            false  thrpt   40  37.600 ± 0.190  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap             true  thrpt   40  30.313 ± 0.167  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap         false  thrpt   40  14.004 ± 0.106  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap          true  thrpt   40  12.960 ± 0.265  ops/us
```

`4.1` but changed to work with the *NoMap benchmarks:
```
Benchmark                                                       (confuseJIT)   Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecodingNoMap                         false  thrpt   40   3.242 ± 0.096  ops/us
QueryStringDecoderBenchmark.mixedDecodingNoMap                          true  thrpt   40   2.543 ± 0.108  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                            false  thrpt   40  15.523 ± 0.135  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                             true  thrpt   40  17.761 ± 0.951  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                   false  thrpt   40   4.318 ± 0.063  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                    true  thrpt   40   3.530 ± 0.070  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                          false  thrpt   40   3.219 ± 0.312  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                           true  thrpt   40   2.810 ± 0.090  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap            false  thrpt   40  38.327 ± 0.335  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap             true  thrpt   40  26.263 ± 1.522  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap         false  thrpt   40   9.730 ± 0.045  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap          true  thrpt   40  10.295 ± 0.674  ops/us
```
In short: the improvement is almost there, in many/most of the cases.
What I've noticed still by inspecting the assembly is that the main decoding method requires some spilling to happen 
which could be saved, but the code is already very complex as it is, so I won't go down that rabbit hole.

I'll now impl the changes suggested by @vietj and we're ready to go.

**Comment by franz1981** [comment]
303b2b2f2c9dacafb1044e0dd7a25e34a0fa0e3b has delivered another interesting speedup in almost all cases but `nonStandardDecodingNoMap` without confusingJIT

```
Benchmark                                                       (confuseJIT)   Mode  Cnt   Score   Error   Units
QueryStringDecoderBenchmark.mixedDecodingNoMap                         false  thrpt   40   3.408 ± 0.116  ops/us
QueryStringDecoderBenchmark.mixedDecodingNoMap                          true  thrpt   40   3.685 ± 0.114  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                            false  thrpt   40  21.339 ± 0.236  ops/us
QueryStringDecoderBenchmark.noDecodingNoMap                             true  thrpt   40  21.085 ± 0.678  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                   false  thrpt   40   4.856 ± 0.178  ops/us
QueryStringDecoderBenchmark.nonStandardDecodingNoMap                    true  thrpt   40   4.667 ± 0.041  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                          false  thrpt   40   4.084 ± 0.065  ops/us
QueryStringDecoderBenchmark.onlyDecodingNoMap                           true  thrpt   40   3.959 ± 0.141  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap            false  thrpt   40  38.688 ± 0.342  ops/us
QueryStringDecoderBenchmark.singleResourceOneParameterNoMap             true  thrpt   40  35.615 ± 0.295  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap         false  thrpt   40  14.442 ± 0.163  ops/us
QueryStringDecoderBenchmark.singleResourceThreeParametersNoMap          true  thrpt   40  13.737 ± 0.512  ops/us
```
this seems worthy especially in case JIT branch profiling is polluted

**Review by chrisvest — APPROVED** [review]

**Comment by franz1981** [comment]
@vietj @chrisvest the only downside of unifying the map/callback (with context) approaches is that it always allocate an empty linked map in case no parameters are found, while on `4.1` it just return an empty (singleton) one.
Any suggestion is appreciated to restore the previous behaviour and/or decide that is not worthy...

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 3]
```diff
@@ -191,16 +187,61 @@ public String path() {
         return path;
     }
 
+    @UnstableApi
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:190]: I think you can remove this annotation as its package-private anyway

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 4]
```diff
@@ -191,16 +187,61 @@ public String path() {
...
+            params = new LinkedHashMap<String, List<String>>();
+            decodeParams(uri, pathEndIdx(), charset, MAP_PARAMETER_COLLECTOR, params, maxParams, semicolonIsNormalChar);
         }
         return params;
     }
 
+    /**
+     * Decodes {@link #uri()} parameters, reporting them in the provided {@code collector}'s map.
+     */
+    public void decodeParameters(Map<String, List<String>> parameters) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:209]: I think renaming the param makes things more clear 

```suggestion
    public void decodeParameters(Map<String, List<String>> out) {
```

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 5]
```diff
@@ -223,70 +264,240 @@ private int pathEndIdx() {
...
+     * would cause
+     * accept("a", "1", map), accept("b", "2", map), accept("c", "3", map)
+     * <p>
+     * to be called.
+     * <p>
+     * Order of calling {@link #accept} is an implementation details users shouldn't rely on, anyway,
+     * and just store/report/filter them assuming random ordering, instead.
+     */
+    public interface ParameterCollector<C> {
+        void accept(String name, String value, C accumulator);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:487]:
```suggestion
        void collect(String name, String value, C accumulator);
```

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 6]
```diff
@@ -338,18 +549,29 @@ private static String decodeComponent(String s, int from, int toExcluded, Charse
...
+        int firstEscaped = getFirstEscaped(s, from, toExcluded, isPath);
         if (firstEscaped == -1) {
             return s.substring(from, toExcluded);
         }
 
+        return decodeEscapedComponent(s, from, toExcluded, charset, isPath, firstEscaped, len);
+    }
+
+    private static int getFirstEscaped(String s, int from, int toExcluded, boolean isPath) {
+        int cutOff = !isPath? '+' : '%';
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:561]: Let's make this easier to read 
```suggestion
        int cutOff = isPath? '%' : '+' ;
```

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 7]
```diff
@@ -223,70 +264,240 @@ private int pathEndIdx() {
...
+        return to;
+    }
+
+    private static final ParameterCollector<Map<String, List<String>>> MAP_PARAMETER_COLLECTOR =
+            new ParameterCollector<Map<String, List<String>>>() {
+                @Override
+                public void accept(String name, String value, Map<String, List<String>> accumulator) {
+                    List<String> values = accumulator.get(name);
+                    if (values == null) {
+                        values = new ArrayList<String>(1);
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:443]: is 1 a good default ? Maybe use 4 so we dont need to expand directly

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 8]
```diff
@@ -223,70 +264,240 @@ private int pathEndIdx() {
...
+    private static int indexOf(String s, int from, int to, int ch) {
+        for (int i = from; i < to; i++) {
+            if (s.charAt(i) == ch) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private static int skipIf(String s, int from, int to, int ch) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:428]:
```suggestion
    private static int skipIf(CharSequence s, int from, int to, int ch) {
```

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 9]
```diff
@@ -223,70 +264,240 @@ private int pathEndIdx() {
...
             }
+            nameStart = valueEndExclusive + 1;
         }
-        addParam(s, nameStart, valueStart, i, params, charset);
-        return params;
     }
 
-    private static boolean addParam(String s, int nameStart, int valueStart, int valueEnd,
-                                    Map<String, List<String>> params, Charset charset) {
+    private static int indexOf(String s, int from, int to, int ch) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:419]:
```suggestion
    private static int indexOf(CharSequence s, int from, int to, int ch) {
```

File: codec-http/src/test/java/io/netty/handler/codec/http/LegacyQueryStringDecoder.java [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+                i += 3;
+            } while (i < toExcluded && s.charAt(i) == '%');
+            i--;
+
+            strBuf.append(new String(buf, 0, bufIdx, charset));
+        }
+        return strBuf.toString();
+    }
+
+    private static int findPathEndIndex(String uri) {
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/LegacyQueryStringDecoder.java:383]:
```suggestion
    private static int findPathEndIndex(CharSequence uri) {
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java (hunk 7 above)
Comment (franz1981) [reply codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java:443]: I've no idea - I would expect that query params usually have a single value for each param, most of the time, rather then more, but I have not much exp in it

**Comment by normanmaurer** [comment]
@franz1981 what is the status of that ?

**Comment by franz1981** [comment]
Sorry @normanmaurer I have parked this due to others which required more attention; I will change it again to unify some of the existing code paths
